### PR TITLE
Simplify key-based thread locals

### PR DESCRIPTION
**Pull request by joboet**
This PR simplifies key-based thread-locals by:
* unifying the macro expansion of `const` and non-`const` initializers
* reducing the amount of code in the expansion
* simply reallocating on recursive initialization instead of going through `LazyKeyInner`
* replacing `catch_unwind` with the shared `abort_on_dtor_unwind`

It does not change the initialization behaviour described in #110897.

### CODE REVIEW

**Comment by cuviper** [comment]
r? @m-ou-se 

^ since you hopefully remember more about this code from #110897.

File: library/std/src/sys/thread_local/os_local.rs [hunk 1]
```diff
@@ -85,78 +48,78 @@ pub macro thread_local_inner {
...
+        unsafe {
+            self.os.set(ptr as *mut u8);
+        }
+        if !old.is_null() {
+            // If the variable was recursively initialized, drop the old value.
+            // SAFETY: We cannot be inside a `LocalKey::with` scope, as the
+            // initializer has already returned and the next scope only starts
+            // after we return the pointer. Therefore, there can be no references
+            // to the old value.
+            drop(unsafe { Box::from_raw(old) });
```
Comment (m-ou-se) [inline library/std/src/sys/thread_local/os_local.rs:118]: I think it'd be good to swap this around (such that it behaves like OnceCell) in a future PR. (That PR will need an FCP and perhaps a crater run though.)

**Comment by m-ou-se** [comment]
This looks great.

@bors r+

**Comment by joboet** [comment]
The failing UI tests only check internal perma-unstable implementations, so I think they are fine to remove.

@rustbot ready

**Comment by joboet** [comment]
r? libs

**Comment by joboet** [comment]
Rebased to resolve the conflict with the native TLS rewrite. I've also added a second commit that makes the `#[allow(dead_code)]` more local and removed the unnecessary imports that are now detected.

**Pull request merged**